lib/ukrand/random.c

```diff
@@ -0,0 +1,82 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <uk/random.h>
+#include <string.h>
+
+static int uk_hwrand_init(struct uk_init_ctx *ictx __unused)
+{
+	return ukarch_random_init();
+}
+uk_early_initcall(uk_hwrand_init, 0x0);
+
+#if CONFIG_LIBUKRAND_HW
```

Comment:
I think that according to your `Config.uk`, `CONFIG_LIBUKRAND_HW` and `CONFIG_LIBUKRAND_CHACHA` would be able to coexist. However this cannot be the case as you would get redeclaration errors. You could go with
```C
#if
...
#elif
#else
#error "Error about config not being selected"
#endif
```

Thus, I would suggest using `CONFIG_LIBUKRAND_CHACHA` as the default one in the configuration (`default y`) since it is a software based, architecture agnostic, solution.

Furthermore, I would add a dependency for each of the options that asks that they depend on each other not being enabled. However this is not exactly nice as selecting one means making the other not visible in the menu, so I would advise using a KConfig hint/comment here:
```
comment "Hint: Hardware randomization depends on ChaCha randomization being disabled"
	depends on CONFIG_LIBUKRAND_CHACHA

comment "Hint: ChaCha randomization depends on Hardware randomization being disabled"
	depends on CONFIG_LIBUKRAND_HW
```

However this still feels awkward. I feel like we should still be able to combine them somehow if we wish to add an additional layer of randomization. That being said, what is the reason this design decision was taken w.r.t. having the two mutually exclusive? I am sorry haven't attended much of the meetings, I do not have enough background about this. Do you happen to know @michpappas ?

Reply:
> CONFIG_LIBUKRAND_HW is always used - either to provide random numbers directly or to seed CONFIG_LIBUKRAND_CHACHA.

Are you sure? I admit I misinterpreted it at first and did not see it's a `choice`. Being a "choice" means they are indeed mutually exclusive, right?. Where does HW seed ChaCha? From what I can see, ChaCha is purely software based in the current state of the PR.

Reply:
Sorry, I misread your comment. It is true that the constants `CONFIG_LIBUKRAND_HW` and `CONFIG_LIBUKRAND_CHACHA` are mutually exclusive. I will use the `if-elif-else` syntax.

> CONFIG_LIBUKRAND_HW is always used 

What I meant is that hardware randomness is always used.

> ChaCha is purely software based in the current state of the PR.

Check [`uk_chacha_init`](https://github.com/unikraft/unikraft/pull/1008/files#diff-2d44fcec5b2b3006cefe77e90a58d3ff5677c279ca6ac8284531bd8776524d25R150), it uses `ukarch_random_seed_u32` to generate the initialization vector.